src/library_pthread.js

```diff
@@ -1069,6 +1069,87 @@ var LibraryPThread = {
 #endif
   },
 
+#if MAIN_MODULE
+  $promiseMap: "new Map();",
+  $nextPromiseId: 1,
+
+  // Create a new promise that can be resolved or rejected by passing
+  // a unique ID to emscripten_promise_resolve/emscripten_promise_reject
+  // TODO(sbc): Should be factor this out and make this API public?
+  $newNativePromise__deps: ['$promiseMap', '$nextPromiseId'],
+  $newNativePromise: function(func, args) {
+    return new Promise((resolve, reject) => {
+      var promiseId = nextPromiseId;
+      nextPromiseId += 1;
+      promiseMap.set(promiseId, {resolve, reject});
+      // Native promise function take promise ID as last argument
+      args.push(promiseId);
+      func.apply(null, args);
+    });
+  },
+
+  _emscripten_promise_resolve__deps: ['$promiseMap'],
+  _emscripten_promise_resolve__sig: 'vip',
+  _emscripten_promise_resolve: function(id, value) {
+#if RUNTIME_DEBUG
+    err('emscripten_resolve_promise: ' + id);
+#endif
+    assert(promiseMap.has(id));
+    promiseMap.get(id).resolve(value);
+    promiseMap.delete(id);
+  },
+
+  _emscripten_promise_reject__deps: ['$promiseMap'],
+  _emscripten_promise_reject__sig: 'vip',
+  _emscripten_promise_reject: function(id) {
+#if RUNTIME_DEBUG
+    dbg('emscripten_promise_reject: ' + id);
+#endif
+    assert(promiseMap.has(id));
+    promiseMap.get(id).reject();
+  },
+
+  // Called on the main thread to syncronize the code loaded on all threads.
+  // This work happens asyncronously. The `callback` is called once this work
+  // is completely, passing the ctx.
+  // TODO(sbc): Should make a new form of __proxy attribute for JS library
+  // function that run asyncroubly like but block the caller until they are
+  // done.  Perhaps "async_with_ctx"?
+  _emscripten_sync_all_threads__sig: 'viii',
+  _emscripten_sync_all_threads__deps: ['_emscripten_proxy_sync_code', '$newNativePromise'],
+  _emscripten_sync_all_threads: function(caller, callback, ctx) {
```

Comment:
This **sync**hronizes all threads, but does so in an **async** manner. Perhaps it should be `_emscripten_async_all_threads` with an `a`? (I realize that's not valid English, and this is half-jokingly, but still maybe "async" makes sense in the name?)

Reply:
Agree we should have a better name here.. the word "sync" is overloaded very badly.  Can you think of a better word for the "catchup" operation?

Reply:
I would go with `_emscripten_sync_all_threads_async`. It seems clear that `sync` is a verb there and `async` is an adjective.

Reply:
Adding to the confusion is that (AFAICS) sometimes we want to just catch ourselves up, and sometimes we want to catch all the threads up. Maybe it would help if we uniformly referred to this operation as synchronizing library lists* instead of synchronizing threads?
or maybe we could use "reconcile" like the accounting term for that? 

* or whatever we call the list of loaded libraries